flang/lib/Optimizer/Transforms/AffinePromotion.cpp

```diff
@@ -227,6 +199,54 @@ class AffineLoopConversion : public mlir::OpRewritePattern<fir::LoopOp> {
   }
 
 private:
+  std::pair<mlir::AffineForOp, mlir::Value>
+  createAffineFor(fir::LoopOp op, mlir::PatternRewriter &rewriter) const {
+    if (auto constantStep = constantIntegerLike(op.step()))
+      if (constantStep.getValue() > 0)
+        return positiveConstantStep(op, constantStep.getValue(), rewriter);
+    return genericBounds(op, rewriter);
+  }
+  std::pair<mlir::AffineForOp, mlir::Value>
+  positiveConstantStep(fir::LoopOp op, int64_t step,
+                       mlir::PatternRewriter &rewriter) const {
+    auto affineFor = rewriter.create<mlir::AffineForOp>(
+        op.getLoc(), ValueRange(op.lowerBound()),
+        mlir::AffineMap::get(0, 1,
+                             mlir::getAffineSymbolExpr(0, op.getContext())),
+        ValueRange(op.upperBound()),
+        mlir::AffineMap::get(0, 1,
+                             1 + mlir::getAffineSymbolExpr(0, op.getContext())),
+        step);
+    return std::make_pair(affineFor, affineFor.getInductionVar());
+  }
+  std::pair<mlir::AffineForOp, mlir::Value>
+  genericBounds(fir::LoopOp op, mlir::PatternRewriter &rewriter) const {
+    auto lowerBound = mlir::getAffineSymbolExpr(0, op.getContext());
+    auto upperBound = mlir::getAffineSymbolExpr(1, op.getContext());
+    auto step = mlir::getAffineSymbolExpr(2, op.getContext());
+    mlir::AffineMap upperBoundMap =
+        mlir::AffineMap::get(0, 3, (upperBound - lowerBound).ceilDiv(step));
```

Comment:
I think the correct from might be `(upperBound - lowerBound).floorDiv(step) + 1`
Fortran defines the number of trips as `(upper − lower +step)/step`
So if you take step = 3, and lower = 0, the number of trips is incremented when upper%3=0. With `(upper).ceilDiv(3)` formula it would increments when upper%3 = 1.

Reply:
Let `n = (upper - lower + step)`. If `n < step` then the number of iterations is `0`. If `n >= k * step and n < (k+1) step` then the number of iterations is `k forall k > 0`.

A semi-open interval could be `i in [0 .. k)` and `k = n floordiv step`.

This interval can be mapped back to the original domain with an affine map of `(i)[lower,step] -> lower + (i * step)`.

Reply:
you are right, I will fix it.